tls/s2n_psk.c

```diff
@@ -278,56 +333,104 @@ S2N_RESULT s2n_finish_psk_extension(struct s2n_connection *conn)
     return S2N_RESULT_OK;
 }
 
-static S2N_RESULT s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac psk_hmac_alg)
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac)
 {
-    switch(psk_hmac_alg) {
+    notnull_check(psk);
+    switch(hmac) {
         case S2N_PSK_HMAC_SHA224:     psk->hmac_alg = S2N_HMAC_SHA224; break;
         case S2N_PSK_HMAC_SHA256:     psk->hmac_alg = S2N_HMAC_SHA256; break;
         case S2N_PSK_HMAC_SHA384:     psk->hmac_alg = S2N_HMAC_SHA384; break;
         default:
-            BAIL(S2N_ERR_HMAC_INVALID_ALGORITHM);
+            S2N_ERROR(S2N_ERR_HMAC_INVALID_ALGORITHM);
+    }
+    return S2N_SUCCESS;
+}
+
+static S2N_CLEANUP_RESULT s2n_psk_parameters_wipe_external_psks(struct s2n_psk_parameters *params)
+{
+    ENSURE_REF(params);
+
+    /* Remove non-external PSKs to avoid wiping them */
+    for (uint32_t i = params->psk_list.len; i > 0; i--) {
+        struct s2n_psk *psk;
+        GUARD_RESULT(s2n_array_get(&params->psk_list, i - 1, (void**)&psk));
+        if (psk->type == S2N_PSK_TYPE_EXTERNAL) {
```

Comment:
To check my understanding here, you are removing the resumption psks from the list because you do not need to deal with their memory, only with external psks do you need to explicitly wipe them?

Reply:
Exactly. We're swapping out the external PSKs, not the resumption PSKs. So regardless of which list we end up using in this API (new, or old bc new was invalid), the resumption PSKs are unaffected and should not be touched.

Reply:
Here's what I keep getting tripped up by. This function removes resumption PSKs from the params list, but it is called "s2n_psk_parameters_wipe_external_psks" which sort of implies that we don't touch the resumption PSKs. Also it feels like this function could be combined with s2n_psk_parameters_wipe()? Why do these functions need to be separate?